pkg/git.go

```diff
@@ -35,16 +45,181 @@ func NewGitPackage(source *spec.GitSource) Interface {
 	}
 }
 
+func downloadGitHubArchive(filepath string, url string) (string, error) {
+	// Get the data
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	color.Cyan("GET %s %d", url, resp.StatusCode)
+
+	// GitHub conveniently uses the commit SHA1 at the ETag
+	// signature for the archive. This is needed when doing `jb update`
+	// to resolve a ref (ie. "master") to a commit SHA1 for the lock file
+	etagValue := resp.Header.Get(http.CanonicalHeaderKey("ETag"))
+	commitShaPattern, _ := regexp.Compile("^\"([0-9a-f]{40})\"$")
+	m := commitShaPattern.FindStringSubmatch(etagValue)
+	if len(m) < 2 {
+		return "", errors.New(fmt.Sprintf("unexpected etag format: %s", etagValue))
+	}
+	commitSha := m[1]
+	defer resp.Body.Close()
+
+	// Create the file
+	out, err := os.Create(filepath)
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+
+	// Write the body to file
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return commitSha, nil
+}
+
+func gzipUntar(dst string, r io.Reader, subDir string) error {
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		return err
+	}
+	defer gzr.Close()
+
+	tr := tar.NewReader(gzr)
+
+	for {
+		header, err := tr.Next()
+		switch {
+		case err == io.EOF:
+			return nil
+
+		case err != nil:
+			return err
+
+		case header == nil:
+			continue
+		}
+
+		// strip the two first components of the path
+		parts := strings.SplitAfterN(header.Name, "/", 2)
+		if len(parts) < 2 {
+			continue
+		}
+		suffix := parts[1]
+		prefix := dst
+
+		// reconstruct the target parh for the archive entry
+		target := filepath.Join(prefix, suffix)
+
+		// if subdir is provided and target is not under it, skip it
+		subDirPath := filepath.Join(prefix, subDir)
+		if subDir != "" && !strings.HasPrefix(target, subDirPath) {
+			continue
+		}
+
+		// check the file type
+		switch header.Typeflag {
+
+		// create directories as needed
+		case tar.TypeDir:
+			if _, err := os.Stat(target); err != nil {
+				if err := os.MkdirAll(target, 0755); err != nil {
```

Comment:
I believe you could use `header.Mode` instead of `0755` here.

Reply:
Yes, done.